src/sensors/zjs_sensor_accel.c

```diff
@@ -0,0 +1,134 @@
+{
+    sensor_handle_t *handle = (sensor_handle_t *)h;
+    jerry_value_t obj = handle->sensor_obj;
+
+    jerry_value_t null_val = jerry_create_null();
+    zjs_set_readonly_property(obj, "x", null_val);
+    zjs_set_readonly_property(obj, "y", null_val);
+    zjs_set_readonly_property(obj, "z", null_val);
+
+    if (jerry_value_has_error_flag(zjs_sensor_stop_sensor(obj))) {
+        zjs_sensor_trigger_error(obj, "SensorError", "stop failed");
+    }
+}
+
+static ZJS_DECL_FUNC(zjs_sensor_constructor)
+{
+    ZJS_VALIDATE_ARGS(Z_OPTIONAL Z_OBJECT);
+
+    double frequency = DEFAULT_SAMPLING_FREQUENCY;
+    jerry_size_t size = SENSOR_MAX_CONTROLLER_NAME_LEN;
+    sensor_controller_t controller;
+
+    if (argc >= 1) {
+        jerry_value_t options = argv[0];
+        ZVAL controller_val = zjs_get_property(options, "controller");
+
+        if (jerry_value_is_string(controller_val)) {
+            zjs_copy_jstring(controller_val, controller.name, &size);
+        } else {
+            snprintf(controller.name, size, BMI160_DEVICE_NAME);
+            DBG_PRINT("controller not set, default to %s\n", controller.name);
+        }
+
+        double option_freq;
+        if (zjs_obj_get_double(options, "frequency", &option_freq)) {
+            // limit frequency to 800
+            if (option_freq < 1 || option_freq > 800) {
+                ERR_PRINT("unsupported frequency, default to %dHz\n",
+                          DEFAULT_SAMPLING_FREQUENCY);
+            } else {
+                frequency = option_freq;
+            }
+        } else {
+            DBG_PRINT("frequency not found, default to %dHz\n",
+                      DEFAULT_SAMPLING_FREQUENCY);
+        }
+    }
+
+    jerry_value_t sensor_obj = zjs_sensor_create(g_instance,
+                                                 SENSOR_CHAN_ACCEL_XYZ,
+                                                 &controller,
+                                                 frequency,
+                                                 onchange,
+                                                 onstart,
+                                                 onstop);
+
+    if (jerry_value_has_error_flag(sensor_obj)) {
+        return sensor_obj;
+    }
+
+    // initialize readings to null
+    ZVAL null_val = jerry_create_null();
+    zjs_set_readonly_property(sensor_obj, "x", null_val);
+    zjs_set_readonly_property(sensor_obj, "y", null_val);
+    zjs_set_readonly_property(sensor_obj, "z", null_val);
+
+    return sensor_obj;
+}
+
+sensor_instance_t *zjs_sensor_accel_init()
+{
+    if (g_instance)
+        return g_instance;
+
+    ZVAL global_obj = jerry_get_global_object();
+    zjs_obj_add_function(global_obj, zjs_sensor_constructor, "Accelerometer");
+    g_instance = zjs_malloc(sizeof(sensor_instance_t));
+    if (!g_instance) {
+        ERR_PRINT("failed to allocate sensor_instance_t\n");
+        return NULL;
+    }
+    memset(g_instance, 0, sizeof(sensor_instance_t));
+    return g_instance;
+}
```

Comment:
I don't like so much code being duplicated in four files. Maybe some of this boilerplate at least could be implemented by a helper function in sensor.c and you call e.g. zjs_sensor_generic_init("Accelerometer") or something.

There are only 14 lines of differences between accel.c and gyro.c, only two of them are "real".

I guess this doesn't have to be dealt with right now, but I don't think we should leave it like this.